### PR TITLE
Copter waypoint radius for each Commands in mission

### PR DESCRIPTION
**Pull request by rahulOCR**
In Copter We sometimes need Waypoint radius to be different for each waypoint because of different area. So I have make small changes in it to accept WP_radius from P1 column as an input from user for different waypoints. Applicable only for MAV_CMD_NAV_WAYPOINT.

### CODE REVIEW

**Review by khancyr — CHANGES_REQUESTED** [review]
Hello,

Unfortunately, we cannot  do this as P1 is for  hold time see https://mavlink.io/en/messages/common.html#MAV_CMD_NAV_WAYPOINT

So this would break every configurations.

**Comment by IamPete1** [comment]
I guess the existing param 2 "Accept Radius" could be used for this. However I suspect SCurves will not work correctly with per waypoint radius currently.

**Comment by rahulOCR** [comment]
You are right, P1 is used  for  hold time. To overcome that I have created another field taking input from P2 as that was empty field. But I Don't know why after uploading mission and reading from EEPROM, it shows random weird data in the mission table.
this in /libraries/AP_Mission/AP_Mission.h
`// command structure
    struct Mission_Command {
        uint16_t index;             // this commands position in the command list
        uint16_t id;                // mavlink command id
        uint16_t p1;                // general purpose parameter 1
        uint16_t pp2;               // general purpose parameter 2 for wp_radius
        Content content;`

and in /Arducopter/mode_auto.cpp
`case MAV_CMD_NAV_WAYPOINT:                  // 16  Navigate to Waypoint
        do_nav_wp(cmd);
        wp_nav->set_wp_radius(float(cmd.pp2 * 100));
    
        gcs().send_text(MAV_SEVERITY_CRITICAL,"Wavpoint_radius = %d",int(wp_nav->get_wp_radius_cm()/100));
        break;`

**Comment by IamPete1** [comment]
We don't store the full command, only a reduced size (16 bytes I think), so when you have a location you only get one param. The union for different command types allows to give useful names for various things. Mission storage is a little tricky, the read from and write to storage functions in AP_Mission are the best place to look.

File: ArduCopter/mode_auto.cpp [hunk 1]
```diff
@@ -600,6 +600,9 @@ bool ModeAuto::start_command(const AP_Mission::Mission_Command& cmd)
 
     case MAV_CMD_NAV_WAYPOINT:                  // 16  Navigate to Waypoint
         do_nav_wp(cmd);
+        wp_nav->set_wp_radius(float(cmd.p1 * 100));
+    
+        gcs().send_text(MAV_SEVERITY_CRITICAL,"Wavpoint_radius = %d",int(wp_nav->get_wp_radius_cm()/100));
```
Comment (muramura) [inline ArduCopter/mode_auto.cpp:605]: I think this message is at the notification or information level.

Comment (rahulOCR) [reply ArduCopter/mode_auto.cpp:605]: Yes this is information level message which is just for debugging purpose to exactly know WP_radius at that waypoint

File: libraries/AC_WPNav/AC_WPNav.h [hunk 2]
```diff
@@ -19,6 +19,15 @@ class AC_WPNav
     /// Constructor
     AC_WPNav(const AP_InertialNav& inav, const AP_AHRS_View& ahrs, AC_PosControl& pos_control, const AC_AttitudeControl& attitude_control);
 
+    // changed getter and setter of waypoint radius //////////////////////////////////////////////////////////////
+
+    void set_wp_radius(float radius) {_wp_radius_cm = radius;}
```
Comment (muramura) [inline libraries/AC_WPNav/AC_WPNav.h:24]: The WP radius is at least 5 cm.
Config parameters are set using the .set method.

    // check _wp_radius_cm is reasonable
    _wp_radius_cm.set(MAX(_wp_radius_cm, WPNAV_WP_RADIUS_MIN));

**Comment by rahulOCR** [comment]
Just make a fine adjustment to use Hold time and WP_radius both using a "TOGGLE" parameter

**Comment by mustafa-gokce** [comment]
I was working something like this but did not get much attention:
https://github.com/ArduPilot/ardupilot/pull/20173

**Comment by rahulOCR** [comment]
@mustafa-gokce Got your point, Thanks for it. I Think behavior is what matters more when it take turns than WP_Radius.

File: libraries/AP_Mission/AP_Mission.h [hunk 3]
```diff
@@ -333,6 +333,7 @@ class AP_Mission
         uint16_t index;             // this commands position in the command list
         uint16_t id;                // mavlink command id
         uint16_t p1;                // general purpose parameter 1
+        uint16_t pp2;               // general purpose parameter 2 for wp_radius
```
Comment (muramura) [inline libraries/AP_Mission/AP_Mission.h:336]: What does "PP" stand for?
I think it is P2.

**Review by rahulOCR — COMMENTED** [review]
Just modify, hold time on high byte and way-point radius in low byte

**Pull request closed without merging**